sw/device/lib/dif/dif_alert_handler.h

```diff
@@ -0,0 +1,649 @@
+ * as part of its policy.
+ */
+typedef uint32_t dif_alert_handler_signal_t;
+
+/**
+ * An alert class state
+ *
+ * Every class begins in an idle state. When sufficient alerts have
+ * accumulated, the following procedure occurs:
+ * - The alert class IRQ is fired to the processor.
+ * - If the processor fails to clear the IRQ within a set timeout,
+ *   escalation phase 0 begins, potentially triggering an escalation signal.
+ * - After some time, phase 1 begins, and so on.
+ * - Once phase 3's duration ends, the class enters the "terminal" state, which
+ *   cannot be exited except by reset.
+ *
+ * At any point, software may end the escalation protocol by calling
+ * `dif_alert_handler_phase_escalation_clear()` (unless clearing is disabled).
+ * Succeessfully calling this function, or clearing the IRQ on time, will reset
+ * back to the idle state.
+ */
+typedef enum dif_alert_handler_class_state {
+  /**
+   * The initial, idle state.
+   */
+  kDifAlertHandlerClassStateIdle,
+  /**
+   * The "timeout" state, that is, the IRQ has been fired and the clock is
+   * ticking for the processor to handle the alert.
+   */
+  kDifAlertHandlerClassStateTimeout,
+
+  /**
+   * The zeroth escalation phase.
+   */
+  kDifAlertHandlerClassStatePhase0,
+  /**
+   * The first escalation phase.
+   */
+  kDifAlertHandlerClassStatePhase1,
+  /**
+   * The second escalation phase.
+   */
+  kDifAlertHandlerClassStatePhase2,
+  /**
+   * The third escalation phase.
+   */
+  kDifAlertHandlerClassStatePhase3,
+
+  /**
+   * The terminal state. Most configurations will never reach this state, since
+   * one of the previous phases will use an escalation signal to reset the
+   * device.
+   */
+  kDifAlertHandlerClassStateTerminal,
+} dif_alert_handler_class_state_t;
+
+/**
+ * The result of an operation.
+ */
+// Open question: basically every function in this DIF can fail due to the
+// registers being locked. Is that a thing we want to include in this error?
+typedef enum dif_alert_handler_result {
+  kDifAlertHandlerOk,
+  kDifAlertHandlerBadArg,
+  kDifAlertHandlerError,
+} dif_alert_handler_result_t;
+
+/**
+ * Alert handler configuration data.
+ *
+ * See `dif_alert_handler_init()`.
+ */
+typedef struct dif_alert_handler_config {
+  /**
+   * The configured number of alerts.
+   *
+   * This value is fixed by the hardware, but not known to this library.
+   */
+  uint32_t alert_count;
+  /**
+   * The configured number of escalation signals.
+   *
+   * This value is fixed by the hardware, but not known to this library.
+   */
+  // It's actually fixed at 4 right now but this is likely to become
+  // configurable too.
+  uint32_t signal_count;
+} dif_alert_handler_config_t;
+
+/**
+ * An alert handler handle.
+ *
+ * This holds state for accessing alart handler registers and should be treated
+ * as an opaque struct.
+ */
+typedef struct dif_alert_handler {
+  mmio_region_t base_addr;
+  dif_alert_handler_config_t config;
```

Comment:
I thought the consensus is that the user of a DIF does state tracking, I vaguely remember my suggestion to keep config in the DIF handler was dismissed in previous reviews.

Reply:
The configuration in this case is not used to "configure" the hardware itself, but is a description of the parametric portions of the hardware. Compare with the rv_timer DIF.

Reply:
`rv_timer` currently uses `config` for this purpose. Let's not introduce a new convention from whole-cloth.